File: api/schemas/debate.py
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, ConfigDict, Field, field_validator
from typing import Optional
from src.debate_enums import DebatePhase, Speaker


class DebateCreateRequest(BaseModel):
    topic: str = Field(..., min_length=1, max_length=500)
    pro_style: str = "passionate"
    con_style: str = "passionate"

    @field_validator("topic")
    @classmethod
    def topic_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("topic must not be blank")
        return v.strip()


class DebateCreateResponse(BaseModel):
    debate_id: str
    topic: str
    pro_style: str
    con_style: str


class StyleInfo(BaseModel):
    name: str
    description: str


class StylesResponse(BaseModel):
    styles: list[StyleInfo]


class DebateSummary(BaseModel):
    """A persisted debate as shown in the 'past debates' list.

    Intentionally omits the (potentially large) transcript and scoreboard — the
    list view only needs the headline facts. ``from_attributes`` lets FastAPI
    build this straight from the ``Debate`` ORM row, including its
    ``message_count`` property.
    """
    model_config = ConfigDict(from_attributes=True)

    id: str
    topic: str
    pro_style: str
    con_style: str
    winner: Optional[str] = None
    message_count: int
    created_at: datetime
    completed_at: datetime


class DebateDetail(DebateSummary):
    """A persisted debate with its full transcript and structured scoreboard."""
    transcript: list[dict]
    argument_scores: Optional[dict] = None


# WebSocket message types
class WSMessageType(str, Enum):
    DEBATE_STARTED = "debate_started"
    PHASE_CHANGE = "phase_change"
    MESSAGE_START = "message_start"
    MESSAGE_CHUNK = "message_chunk"
    MESSAGE_COMPLETE = "message_complete"
    VOTE_REQUIRED = "vote_required"
    VOTE_RECEIVED = "vote_received"
    ARGUMENT_SCORES = "argument_scores"
    DEBATE_COMPLETE = "debate_complete"
    ERROR = "error"


class WSMessage(BaseModel):
    type: WSMessageType
    debate_id: str
    data: dict

File: frontend/src/types/debate.ts
export type DebatePhase =
  | 'introduction'
  | 'opening_pro'
  | 'opening_con'
  | 'rebuttal'
  | 'closing_pro'
  | 'closing_con'
  | 'verdict'
  | 'scoring'
  | 'finished';

export type Speaker = 'PRO' | 'CON' | 'MODERATOR' | 'JUDGE' | 'AUDIENCE' | 'SCORING';

export interface StyleInfo {
  name: string;
  description: string;
}

export interface DebateMessage {
  speaker: Speaker;
  content: string;
  label?: string;
  phase: DebatePhase;
}

// Structured judge scoring (mirrors src/scoring.py DebateScores).
export interface ArgumentScore {
  summary: string;
  score: number;
  reason: string;
}

export interface DebateScores {
  pro_arguments: ArgumentScore[];
  con_arguments: ArgumentScore[];
  pro_average: number;
  con_average: number;
  winner: 'PRO' | 'CON' | 'TIE';
  strongest_argument: string;
  weakest_argument: string;
}

export interface DebateConfig {
  topic: string;
  proStyle: string;
  conStyle: string;
}

export interface DebateSession {
  debateId: string;
  topic: string;
  proStyle: string;
  conStyle: string;
}

// WebSocket message types
export type WSMessageType =
  | 'debate_started'
  | 'phase_change'
  | 'message_start'
  | 'message_chunk'
  | 'message_complete'
  | 'vote_required'
  | 'vote_received'
  | 'argument_scores'
  | 'debate_complete'
  | 'error';

export interface WSMessage {
  type: WSMessageType;
  debate_id: string;
  data: Record<string, unknown>;
}

export interface DebateTranscriptEntry {
  speaker: string;
  content: string;
  phase: string;
}

// Persisted debates (GET /api/debates and /api/debates/{id}). Field names are
// snake_case to match the API payload directly.
export interface PastDebateSummary {
  id: string;
  topic: string;
  pro_style: string;
  con_style: string;
  winner: 'PRO' | 'CON' | 'TIE' | null;
  message_count: number;
  created_at: string;
  completed_at: string;
}

export interface PastDebateDetail extends PastDebateSummary {
  transcript: DebateTranscriptEntry[];
  argument_scores: DebateScores | null;
}
